Add fusion enabled apis

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#64429 Add fusion enabled apis**
* #64428 update optimize_for_inference docs

Differential Revision: [D30740897](https://our.internmc.facebook.com/intern/diff/D30740897)

Commit:
Add fusion enabled apis

[ghstack-poisoned]

ZolotukhinM commented on torch/jit/__init__.py:
```diff
@@ -197,6 +198,34 @@ def _hide_source_ranges() -> Iterator[None]:
+        torch._C._jit_override_can_fuse_on_gpu(enabled)
+    else:
+        assert device in ["cpu", "cuda"], "Device-specific fusion must be equal to 'cpu' or 'cuda' if not None"
+        if device == "cpu":
+            torch._C._jit_override_can_fuse_on_gpu(enabled)
+        else:
+            torch._C._jit_override_can_fuse_on_cpu(enabled)
```
If-else statements seem swapped.

ZolotukhinM reviewed (approved):
Nice!

bertmaher reviewed (approved):
Awesome, thank you!

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30740897).

Commit:
Update on "Add fusion enabled apis"


Differential Revision: [D30740897](https://our.internmc.facebook.com/intern/diff/D30740897)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30740897).
